service/matching/tasklist/matcher.go

```diff
@@ -445,32 +438,17 @@ func (tm *TaskMatcher) fwdrAddReqTokenC() <-chan *ForwarderReqToken {
 	return tm.fwdr.AddReqTokenC()
 }
 
-func (tm *TaskMatcher) ratelimit(ctx context.Context) (*rate.Reservation, error) {
-	select {
-	case <-ctx.Done():
-		return nil, ctx.Err()
-	default:
-	}
-
-	deadline, ok := ctx.Deadline()
-	if !ok {
-		if err := tm.limiter.Wait(ctx); err != nil {
-			return nil, err
-		}
-		return nil, nil
+func (tm *TaskMatcher) ratelimit(ctx context.Context) error {
+	err := tm.limiter.Wait(ctx)
+	if errors.Is(err, clock.ErrCannotWait) {
+		// "err != ctx.Err()" may also be correct, as that would mean "gave up due to context".
+		//
+		// in this branch: either the request would wait longer than ctx's timeout,
+		// or the limiter's config does not allow any operations at all (burst 0).
+		// in either case, this is returned immediately.
+		return ErrTasklistThrottled
 	}
-
-	rsv := tm.limiter.Reserve()
-	// If we have to wait too long for reservation, give up and return
-	if !rsv.OK() || rsv.Delay() > time.Until(deadline) {
-		if rsv.OK() { // if we were indeed given a reservation, return it before we bail out
-			rsv.Cancel()
-		}
-		return nil, ErrTasklistThrottled
```

Comment:
"throttled" == "giving up immediately because there is not enough time" feels somewhat misleading, but I suppose it works.

anyway, this is now returned in all cases where wait does not wait, and returns immediately.  since burst==0 seems probably not possible / things would be horribly broken, that should cover this exactly.